### PR TITLE
Support day of year (`%j`) in `Time` parsers

### PR DESCRIPTION
**Pull request by HertzDevil**
This includes `Time::Format#parse` and everything that depends on it, such as `Time.parse` and JSON deserialization.

Day of year has higher priority over month + day, but lower priority over week date (`%G` `%V` `%u`). The latter is unlikely to cause problems, since day of year and week date in the same string is very unlikely to happen.

### CODE REVIEW

File: spec/std/time/format_spec.cr [hunk 1]
```diff
@@ -522,6 +522,37 @@ describe Time::Format do
...
+    parse_time("2004-200", "%Y-%j").should eq(Time.utc(2004, 7, 18))
+    parse_time("2004-365", "%Y-%j").should eq(Time.utc(2004, 12, 30))
+    parse_time("2004-366", "%Y-%j").should eq(Time.utc(2004, 12, 31))
+
+    expect_raises(Time::Format::Error, "Invalid day of year") do
+      parse_time("2006-000", "%Y-%j")
+    end
+    expect_raises(Time::Format::Error, "Invalid day of year") do
+      parse_time("2004-000", "%Y-%j")
+    end
```
Comment (straight-shoota) [inline spec/std/time/format_spec.cr:547]: Is the day of year value duplicated on purpose?

**Review by straight-shoota — APPROVED** [review]

File: spec/std/time/format_spec.cr (hunk 1 above)
Comment (oprypin) [reply spec/std/time/format_spec.cr:547]: bump regarding this question

Comment (HertzDevil) [reply spec/std/time/format_spec.cr:547]: One is a leap year, other isn't.

**Review by beta-ziliani — APPROVED** [review]
🚀

**Pull request merged**